vllm/v1/worker/gpu_model_runner.py

```diff
@@ -1176,9 +1176,9 @@ def execute_model(
                     input_ids, mm_embeds)
             else:
                 inputs_embeds = self.model.get_input_embeddings(input_ids)
-            # TODO(woosuk): Avoid the copy. Optimize.
-            self.inputs_embeds[:num_scheduled_tokens].copy_(inputs_embeds)
-            inputs_embeds = self.inputs_embeds[:num_input_tokens]
+            if num_scheduled_tokens != num_input_tokens:
```

Comment:
I didn't get this part, why we can skip the copy if num_scheduled_tokens == num_input_tokens. Could you elaborate a bit?

Reply:
You're right, I should've elaborated on this in the PR description. Here's my understanding of the code:

`num_input_tokens` is computed here:
https://github.com/vllm-project/vllm/blob/432ec9926ebd3ce826f0d49df0a2a5ae3cc81ec0/vllm/v1/worker/gpu_model_runner.py#L1145-L1164
So this code simply pads `input_embeds` to `num_input_tokens` if necessary but does it via `self.inputs_embeds` instead of `torch.nn.functional.pad` probably because of performance reasons.
This means if `num_scheduled_tokens == num_input_tokens` and `self.inputs_embeds` is only used needed [dummy input](https://github.com/vllm-project/vllm/blob/432ec9926ebd3ce826f0d49df0a2a5ae3cc81ec0/vllm/v1/worker/gpu_model_runner.py#L1728) we don't actually need to do any padding and therefor can get rid of this copy.

Reply:
Oh interesting, I wasn't aware. I didn't see any failures in my tests, sorry about that.
For the future, how could I reproduce such issues related to cuda graph capture?